### ISSUE TITLE
Investigate errors in SQLAlchemy 1.3

### ISSUE THREAD

**Issue by jeancochrane** [jeancochrane/pytest-flask-sqlalchemy#14]
After bringing in #6, we've had reports of unreliability with some tests:

> @charness thank you for the fix, I couldn't trace this down for a long time!
>
> However, it introduced some flakiness in our test suite for several tests that do a commit and then check if the attribute changes. Have you had anything like it?
>
> The code is similar to this:

```python
def test_should_confirm_account(db_session, app_client):
    user = UserFactory()
    token = user.generate_token_and_save_timestamp()

    db_session.add(user)
    # First commit or flush in some cases
    db_session.commit()

    # One more commit here:
    response = app_client.post("/account/confirm", json={"token": token})
    assert response.status_code == 200

    # Fails with:
    # Instance <User at 0x108bf1d30> is not bound to a Session; attribute refresh operation cannot proceed (Background on this error at: http://sqlalche.me/e/bhk3 
    assert user.is_confirmed()
```

Check to see if this example can be reproduced, and if so, investigate the root cause of the problem.

**Edit 3/12**: This appears to be a problem with SQLAlchemy 1.3. Pinning your SQLAlchemy version to `<1.3` should fix any errors in the meantime while we figure out what exactly is going on.

**Comment by charness** [comment]
The material found via the link in the error message seems to use _transaction_ and _session_ interchangeably, though they are distinct in SQLAlchemy. Indeed, here the session is the same, but it's a new transaction.
> **Don’t close the session prematurely** - Often, applications will close out a transaction before passing off related objects to some other system which then fails due to this error. 

In comparable code I've tested in the past, explicitly reloading the user object by query (e.g., query by id or other key) after the commit would more closely match actual access patterns and should avoid the error.

**Comment by tjwaterman99** [comment]
I also started having some issues that I think might be related.

SQLAlchemy is having trouble committing a new object to the database:

```
sqlalchemy.orm.exc.FlushError: Instance <User at 0x7fee2d44cd30> has a NULL identity key.  

If this is an auto-generated value, check that the database table allows generation of new primary key values, and that the mapped Column object is configured to expect these generated values.  Ensure also that this flush() is not occurring at an inappropriate time, such as within a load() event
```

Here's my `conftest.py`

```python
@pytest.fixture(scope='session')
def app():
    app = create_app()
    with app.app_context():
        yield app


@pytest.fixture(scope='session')
def client(app):
    return app.test_client()


@pytest.fixture(scope='session')
def _db(app):
    return app.extensions['sqlalchemy'].db


@pytest.fixture
def test_user(db_session):
    test_user = User(
        email="test@my.app",
        password="fakepassword"
    )
    db_session.add(test_user)
    db_session.commit()
    return test_user
```

I think it could be unrelated, but I don't see any releases in my other libraries, and this set up was working before `1.0.1`.

**Comment by krispharper** [comment]
Hmm. I'm also seeing this issue, but I think it's unrelated to the release since it also breaks for me on `1.0.0`.

Maybe it's related to a change in one of the dependencies?

**Comment by krispharper** [comment]
Yeah, it appears to be an issue with SQLAlchemy 1.3.0, which was released on Monday. Updating [this line](https://github.com/krispharper/pytest-flask-sqlalchemy/blob/sqlalchemy_version/setup.py#L21) to restrict to version <1.3 fixes the issue for me.

Unfortunately I'm not really sure what the actual problem is.

**Comment by tjwaterman99** [comment]
Thanks @krispharper for pointing that out. Including SQLAlchemy explicitly in my dependencies and using version 1.2.18 also fixes this for me.

**Comment by jeancochrane** [comment]
Yup, I think you all nailed it: the root problem here has something to do with a change introduced in SQLAlchemy 1.3. A number of tests in our test suite start failing in 1.3.0. Updated the title of the issue to reflect this, and added a note to the root issue explaining that pinning SQLAlchemy to <1.3 should fix the issue for temporarily while I work on a more permanent solution. Thanks for all the debugging assistance!

**Comment by jeancochrane** [comment]
I think I've located the source of the problem. Our use of `contextual_connect` is [deprecated as of 1.3](https://docs.sqlalchemy.org/en/latest/changelog/migration_13.html#threadlocal-engine-strategy-deprecated):

https://github.com/jeancochrane/pytest-flask-sqlalchemy/blob/e9ddaabab432f07a233e21ac8326935a6ce2bbd6/pytest_flask_sqlalchemy/fixtures.py#L98

Changing this method to `_contextual_connect` fixes the problem. Two problems I still need to think through before I bring in a fix:

- If the threadlocal strategy is deprecated, do we need to have a mock method for `_contextual_connect` at all? Why is it still being called in the SQLAlchemy source?
- What's the best way to support both 1.2 and 1.3, given this API change?

**Comment by krispharper** [comment]
SQLAlchemy does [include its version](https://github.com/zzzeek/sqlalchemy/blob/master/lib/sqlalchemy/__init__.py#L125), so a (somewhat hacky) way to support both would be to use `sqlalchemy.__version__` and only mock for versions less than 1.3.

Alternatively, new versions of pytest-flask-sqlalchemy could support `sqlalchemy>=1.3` going forward.

**Comment by javiersanp** [comment]
In my setting, db_session.commit fails to persist objects when using sqlalchemy >= 1.3 (tested in 1.3.0 and 1.3.1). With 1.2.18 commit persist well. Example code::

    def test_user(db_session):
        u = User(name="a1234123")
        db_session.add(u)
        db_session.commit()
        assert db_session.query(Station).count() == 1

Fails with sqlalchemy >= 1.3, pass with sqlalchemy <1.3

**Comment by killthekitten** [comment]
@jeancochrane confirmed, changing to `_contextual_connect` fixes everything for me (tested on `pytest-flask-sqlalchemy==1.0.0` and `sqlalchemy==1.3.1]).

**Comment by jeancochrane** [comment]
Just published [v1.0.2](https://github.com/jeancochrane/pytest-flask-sqlalchemy/releases/tag/v1.0.2) providing support for SQLAlchemy 1.3. Thanks for the help figuring this out, all!

**Comment by killthekitten** [comment]
@jeancochrane you're amazing!
